File: solutions/scala/024-01-matrix.scala
/**
 * Problem 24: 01 Matrix (LeetCode 542)
 * Difficulty: Med
 * Language: Scala
 */
def updateMatrix(mat: Array[Array[Int]]): Array[Array[Int]] = {
  val rows = mat.length; val cols = mat(0).length
  val inf = rows + cols
  val dist = mat.map(_.map(v => if (v == 0) 0 else inf))

  for (r <- 0 until rows; c <- 0 until cols) {
    if (r > 0) dist(r)(c) = math.min(dist(r)(c), dist(r-1)(c) + 1)
    if (c > 0) dist(r)(c) = math.min(dist(r)(c), dist(r)(c-1) + 1)
  }
  for (r <- (0 until rows).reverse; c <- (0 until cols).reverse) {
    if (r < rows - 1) dist(r)(c) = math.min(dist(r)(c), dist(r+1)(c) + 1)
    if (c < cols - 1) dist(r)(c) = math.min(dist(r)(c), dist(r)(c+1) + 1)
  }
  dist
}

File: solutions/typescript/024-01-matrix.ts
/**
 * Problem 24: 01 Matrix (LeetCode 542)
 * Difficulty: Med
 * Language: TypeScript
 */
function updateMatrix(mat: number[][]): number[][] {
  const rows = mat.length, cols = mat[0].length;
  const INF = rows + cols;

  // Functional BFS with immutable state
  const initial = mat.map((row, r) =>
    row.map((v, c) => (v === 0 ? 0 : INF))
  );

  // Top-left pass
  const afterTL = initial.map((row, r) =>
    row.map((v, c) => {
      const top = r > 0 ? initial[r - 1][c] + 1 : INF;
      const left = c > 0 ? row[c - 1] + 1 : INF;
      return Math.min(v, top, left);
    })
  );

  // Bottom-right pass (fold from bottom-right)
  const result = [...afterTL].reverse().map((row, ri) => {
    const r = rows - 1 - ri;
    return [...row].reverse().map((v, ci) => {
      const c = cols - 1 - ci;
      const bottom = r < rows - 1 ? afterTL[r + 1][c] + 1 : INF;
      const right = c < cols - 1 ? afterTL[r][c + 1] + 1 : INF;
      const minVal = Math.min(v, bottom, right);
      afterTL[r][c] = minVal;
      return minVal;
    }).reverse();
  }).reverse();

  return result;
}
